runtime/src/leader_bank_status.rs

```diff
@@ -0,0 +1,167 @@
+pub struct LeaderBankStatus {
+    /// Current state of the system
+    status: Mutex<Status>,
+    /// Weak reference to the current bank
+    bank: RwLock<Option<(Slot, Weak<Bank>)>>,
+    /// CondVar to notify status changes and waiting
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator:
+/// [Unininitialized] -> [InProgress] -> [Completed] --|
+///                          ^-------------------------|
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// Initial state, no bank, but also not completed yet
+    #[default]
+    Uninitialized,
+    /// Slot is in progress as leader
+    InProgress,
+    /// PoH has reached the end of the slot, and the next bank as leader is not available yet
+    Completed,
+}
+
+impl LeaderBankStatus {
+    /// Set the status to `InProgress` and notify any waiting threads
+    /// if the status was not already `InProgress`.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut status = self.status.lock().unwrap();
+        if matches!(*status, Status::InProgress) {
+            return;
+        }
+
+        *status = Status::InProgress;
+        *self.bank.write().unwrap() = Some((bank.slot(), Arc::downgrade(bank)));
+        drop(status);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `Completed` and notify any waiting threads
+    /// if the status was not already `Completed`
+    /// and the slot is higher than the current slot (sanity check).
+    pub fn set_completed(&self, slot: Slot) {
+        let mut status = self.status.lock().unwrap();
+        if matches!(*status, Status::Completed) {
+            return;
+        }
+
+        if let Some((current_slot, _)) = *self.bank.read().unwrap() {
+            if slot < current_slot {
+                return;
+            }
+        }
+
+        *status = Status::Completed;
+        drop(status);
+
+        self.condvar.notify_all();
+    }
+
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// Returns `None` if the timeout is reached.
+    pub fn wait_for_in_progress(&self, timeout: Duration) -> Option<Weak<Bank>> {
+        let status = self.status.lock().unwrap();
+
+        // Hold status lock until after the weak bank reference is cloned.
+        let (_status, wait_timeout_result) = self
+            .condvar
+            .wait_timeout_while(status, timeout, |status| {
+                matches!(*status, Status::Uninitialized | Status::Completed)
+            })
+            .unwrap();
+
+        (!wait_timeout_result.timed_out())
+            .then(|| self.bank.read().unwrap().as_ref().unwrap().1.clone())
+    }
+
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_next_completed(&self, mut timeout: Duration) -> Option<Slot> {
+        loop {
```

Comment:
why different looping construct is used than `.wait_for_in_progress()`?

Reply:
The naming of the functions is perhaps not clear enough on why these are different.

`wait_for_in_progress` either immediately returns if the status is already `InProgress`, or waits for it to become `InProgress`.

`wait_for_next_completed` (emphasis on **_next_**) explicitly waits for the next notification, regardless of current status.

This is due to my different use-cases for the notifications. I used `wait_for_in_progress` to either get or wait shortly for a bank to process transactions on. I used `wait_for_next_completed` to report metrics at the end of a slot.

Reply:
oh i see.. how about `get_or_wait_for_in_progress` / `wait_for_complete`? imo, behavior difference is best described by qualifying verbs, not nouns.

Reply:
done: 438b7bd0ab739ca185ff2d59bd94abc4450da50b